karapace/kafka_rest_apis/__init__.py

```diff
@@ -40,36 +50,70 @@ def __init__(self, config: Config) -> None:
         super().__init__(config=config)
         self._add_kafka_rest_routes()
         self.serializer = SchemaRegistrySerializer(config=config)
-        self._cluster_metadata = None
-        self._metadata_birth = None
-        self.metadata_max_age = self.config["admin_metadata_max_age"]
-        self.admin_client = None
-        self.admin_lock = None
-        self.metadata_cache = None
-        self.schemas_cache = {}
-        self.consumer_manager = ConsumerManager(config=config)
-        self.init_admin_client()
+        self.proxies: Dict[str, "UserRestProxy"] = {}
+        self._proxy_lock = asyncio.Lock()
+        log.info("REST proxy starting with (delegated authorization=%s)", self.config.get("rest_authorization", False))
+        self._idle_proxy_janitor_task: Optional[asyncio.Task] = None
 
-        self._async_producer_lock = None
-        self._async_producer: Optional[AIOKafkaProducer] = None
+    async def close(self) -> None:
+        if self._idle_proxy_janitor_task is not None:
+            self._idle_proxy_janitor_task.cancel()
+            self._idle_proxy_janitor_task = None
+        async with AsyncExitStack() as stack:
+            stack.push_async_callback(super().close)
+            stack.push_async_callback(self.serializer.close)
 
-        self.app.on_startup.append(self.initialize_asyncio_locks)
+            for proxy in self.proxies.values():
+                stack.push_async_callback(proxy.aclose)
 
-    async def initialize_asyncio_locks(self, app: aiohttp.web.Application):  # pylint: disable=unused-argument
-        """The locks can be initialized when asyncio loop is available"""
-        self.admin_lock = asyncio.Lock()
-        self._async_producer_lock = asyncio.Lock()
+    async def _idle_proxy_janitor(self) -> None:
+        IDLE_PROXY_TIMEOUT = 5 * 60
+        idle_consumer_timeout = self.config.get("consumer_idle_disconnect_timeout", 0)
+        while True:
+            await asyncio.sleep(IDLE_PROXY_TIMEOUT / 2)
+
+            key, proxy = None, None
```

Comment:
Maybe put one more `try` here. If something goes wrong in the code, I don't know what, the janitor will die. It'd be probably better if the exception would be logged and janitor would just sleep and try again.

Reply:
There is now simplification by splitting into two, and all catching try block